### DISCUSSION TITLE
v2.0.0-beta.5

### DISCUSSION THREAD

**Discussion by pelletier** [pelletier/go-toml#718] [Releases]
While this release focused on adding back missing encoder and decoder features, it also comes with a slew of bug fixes, thanks to @moorereason's awesome differential fuzzing work! This is the last planned pre-release dedicated to the core library. Check out the updated readme for [upgrade instructions](https://github.com/pelletier/go-toml/blob/v2/README.md#migrating-from-v1).

The next milestone is to port the existing tooling (tomljson, tomllint, binary releases) to go-toml v2.

Thank you, and happy holidays! 

### What's new
* API: Encoder and Decoder options are chainable by @pelletier in https://github.com/pelletier/go-toml/pull/670
* Decoder: show struct field in type mismatch errors by @pelletier in https://github.com/pelletier/go-toml/pull/684
* Encoder: omitempty flag by @pelletier in https://github.com/pelletier/go-toml/pull/692
* Decoder: allow commas in tags by @pelletier in https://github.com/pelletier/go-toml/pull/693
* Decoder: time allows extra precision by @pelletier in https://github.com/pelletier/go-toml/pull/710
* Encode: add comment struct tag by @pelletier in https://github.com/pelletier/go-toml/pull/711
### Performance
* Decoder: use lookup table to validate ASCII by @pelletier in https://github.com/pelletier/go-toml/pull/654
* Decode: code cleanup for struct cache by @pelletier in https://github.com/pelletier/go-toml/pull/659
### Fixed bugs
* Decode: error on underscore before exponent in floats by @pelletier in https://github.com/pelletier/go-toml/pull/647
* Decode: add missing checks for LocalTime by @pelletier in https://github.com/pelletier/go-toml/pull/650
* Decoder: flag invalid carriage returns in strings by @pelletier in https://github.com/pelletier/go-toml/pull/652
* Decode: correctly handle carriage return in comments by @pelletier in https://github.com/pelletier/go-toml/pull/656
* Decode: prevent comments that look like dates to be accepted by @pelletier in https://github.com/pelletier/go-toml/pull/657
* Decoder: validate inline tables don't redefine keys by @pelletier in https://github.com/pelletier/go-toml/pull/660
* Decoder: validate inline tables in arrays don't redefine keys by @pelletier in https://github.com/pelletier/go-toml/pull/663
* Decode: accept E as exponent notation by @moorereason in https://github.com/pelletier/go-toml/pull/665
* Decoder: prevent duplicates of inline tables by @pelletier in https://github.com/pelletier/go-toml/pull/667
* Decoder: flag invalid carriage returns in literal strings by @moorereason in https://github.com/pelletier/go-toml/pull/673
* Encoder: try to use pointer type TextMarshaler by @pelletier in https://github.com/pelletier/go-toml/pull/679
* Decoder: validate bounds of day and month in dates by @pelletier in https://github.com/pelletier/go-toml/pull/680
* Decoder: fail on unescaped \r not followed by \n by @pelletier in https://github.com/pelletier/go-toml/pull/681
* Decoder: fix panic on table array behind a pointer by @pelletier in https://github.com/pelletier/go-toml/pull/682
* Decoder: flag bad \r in literal multiline strings by @pelletier in https://github.com/pelletier/go-toml/pull/687
* Decoder: check timezones start with +,-,z,Z by @pelletier in https://github.com/pelletier/go-toml/pull/688
* Decoder: check tz has : between hours and minutes by @pelletier in https://github.com/pelletier/go-toml/pull/691
* Decoder: check for invalid characters in timezone by @pelletier in https://github.com/pelletier/go-toml/pull/695
* Decode: restrict timezone offset values by @moorereason in https://github.com/pelletier/go-toml/pull/696
* Decode: ensure signed numbers don't start with an underscore by @moorereason in https://github.com/pelletier/go-toml/pull/698
* Decode: allow maximum seconds value of 60 by @moorereason in https://github.com/pelletier/go-toml/pull/700
* Decode: ensure signed exponents don't start with an underscore by @moorereason in https://github.com/pelletier/go-toml/pull/699
* Decoder: prevent modification of inline tables by @pelletier in https://github.com/pelletier/go-toml/pull/702
* Decoder: disallow modification of existing table by @pelletier in https://github.com/pelletier/go-toml/pull/704
* Decode: fix maximum time offset values by @moorereason in https://github.com/pelletier/go-toml/pull/706
* Decode: allow CRLF as whitespace before line ending backslash by @pelletier in https://github.com/pelletier/go-toml/pull/709
* Decode: fix index panic on integers with just a sign by @moorereason in https://github.com/pelletier/go-toml/pull/716
* Decode: check buffer length before parsing simple key by @moorereason in https://github.com/pelletier/go-toml/pull/717
### Documentation
* Encoder: mention "-" to prevent encoding field in doc by @pelletier in https://github.com/pelletier/go-toml/pull/683
* Readme: document more differences with v1 by @pelletier in https://github.com/pelletier/go-toml/pull/712
* Doc: clarify errors returned by Decode by @pelletier in https://github.com/pelletier/go-toml/pull/713

**Full Changelog**: https://github.com/pelletier/go-toml/compare/v2.0.0-beta.4...v2.0.0-beta.5

<hr /><em>This discussion was created from the release <a href='https://github.com/pelletier/go-toml/releases/tag/v2.0.0.5'>v2.0.0-beta.5</a>.</em>

**Comment by moorereason** [comment]
I agree.  I think I've unearthed the low-hanging fruit as far as fuzzing goes.

FWIW: I've primarily been doing differential fuzzing against [`toml.dart`](https://github.com/just95/toml.dart) since it seems to be a solid parser (I've yet to find any bugs in it).  The other projects I've tried that have a toml-test decoder ([BurntSushi](https://github.com/BurntSushi/toml), [tomlc99](https://github.com/cktan/tomlc99), [toml11](https://github.com/ToruNiina/toml11), and [toml-f](https://github.com/toml-f/toml-f)) all have outstanding bugs that need to be resolved before I can continue fuzzing with them.
